nemo_curator/download/commoncrawl.py

```diff
@@ -71,55 +73,126 @@ def lang_detect(decoded_html):
 def extract_text(
     html,
     stop_words,
-    length_low=70,
-    length_high=200,
-    stopwords_low=0.30,
+    algorithm="justext",
+    length_low=None,
+    length_high=None,
+    stopwords_low=None,
     stopwords_high=0.32,
-    max_link_density=0.2,
-    max_heading_distance=200,
-    no_headings=False,
+    max_link_density=None,
+    max_heading_distance=None,
+    no_headings=None,
     logger=None,
 ):
-    # Segment the HTML into paragraphs
-    try:
-        # Form the DOM tree
-        dom = justext.core.html_to_dom(html)
-        cleaned_dom = justext.core.preprocessor(dom)
-        # Get the paragraphs from the DOM
-        handler = justext.core.ParagraphMaker()
-        lxml.sax.saxify(cleaned_dom, handler)
-    except (lxml.etree.ParserError, ValueError, Exception):
-        # Return nothing when we cannot segment the document
-        if logger is not None:
-            logger.info("Could not segment paragaphs in the document")
-        return
-    paragraphs = handler.paragraphs
-
-    # Context free classification
-    justext.core.classify_paragraphs(
-        paragraphs,
-        stop_words,
-        length_low,
-        length_high,
-        stopwords_low,
-        stopwords_high,
-        max_link_density,
-        no_headings,
-    )
+    if algorithm.lower() == "justext":
+        if length_low is None:
```

Comment:
I don't want to over engineer this simple task, but given how many arguments we want to expose I think it's worth refactoring this into some classes even if they're a bit heavier.

I imagine something like this could work

```python
class HTMLExtractionAlgorithm(ABC):
  @abstractmethod
  def extract_text(html, stop_words):
    pass

class JusTextExtraction(HTMLExtractionAlgorithm):
  def __init__(self, length_low=70, ...):
     self.length_low = length_low
     # Etc...

  def extract_text(html, stop_words):
     # Segment the HTML into paragraphs
    try:
        # Form the DOM tree
        dom = justext.core.html_to_dom(html)
    # Etc...
    except:
      # Etc...

class Resiliparse(HTMLExtractionAlgorithm):
  # Etc...
```

Then, when the user calls `download_common_crawl` they could supply something like this:
```
my_algorithm = Resiliparse(...)
download_common_crawl('/path/to/output', '2020-01', '2020-03', algorithm=my_algorithm)
```
Keeping the default as the default `JusTextExtraction()`

My main reasons for thinking this is preferred are:
1. This gets rid of the weird `None` but still default arguments.
    1. The defaults will be easily visible in each class' constructor.
    1. No warnings are needed.
1. It makes clear which arguments are necessary for which algorithm.
    1. It also decouples the defaults. I see that `stopwords_high` is used for both parsers, but it's not immediately obvious that both algorithms should share the same default to me.
1. It keeps each algorithm's logic isolated from each other.

I'd like to hear your opinion on this too if you think there's a better way to handle this.

Reply:
I think the classes are a good idea. My only comment is that I think for whichever one we end up setting as the default, the user shouldn't have to touch any of those parameters (including having to set the algorithm itself). I can still create the class logic and then allow `algorithm` to default to `JusTextExtraction()`.

So while
```
my_algorithm = ResiliparseExtraction(...)
download_common_crawl('/path/to/output', '2020-01', '2020-03', algorithm=my_algorithm)
```
is an option, we should have the default as:
```
download_common_crawl('/path/to/output', '2020-01', '2020-03')  # defaults with JusTextExtraction()
```

Reply:
Totally agree